packages/timeline-state-resolver/src/integrations/vmix/index.ts

```diff
@@ -1454,3 +1432,127 @@ export interface VMixAudioChannel {
 	meterF2: number
 	headphonesVolume: number
 }
+
+/**
+ * Handles pre-loading of inputs.
+ */
+class VMixInputHandler extends EventEmitter {
+	// how long to let an input linger before removing it
+	private TTL = 1000
+
+	private _inputs: Map<
+		string,
+		{
+			type: string
+			name: string
+			added: boolean
+			removeTime?: number
+		}
+	> = new Map()
+
+	private _updateTimeout: NodeJS.Timeout | undefined = undefined
+
+	constructor(private _vmix: VMixDevice) {
+		super()
+	}
+
+	public addInput(key: string, type: string, name: string) {
+		let existing = this._inputs.get(key)
+		if (!existing) {
+			existing = {
+				type,
+				name,
+				added: false,
+			}
+			this._inputs.set(key, existing)
+		}
+		delete existing.removeTime
+
+		this._triggerUpdate()
+	}
+
+	/** Mark input for removal */
+	public removeInput(time: number, key: string) {
+		const existing = this._inputs.get(key)
+		if (existing) {
+			existing.removeTime = Math.min(existing.removeTime ?? Infinity, time + this.TTL)
+
+			this._triggerUpdate()
+		}
+	}
+
+	private _triggerUpdate() {
+		if (this._updateTimeout) {
+			clearTimeout(this._updateTimeout)
+			this._updateTimeout = undefined
+		}
+		setImmediate(() => {
+			try {
+				this._update()
+			} catch (e) {
+				this.emit('error', e)
+			}
+		})
+	}
+	private _update() {
+		const commands: Array<VMixStateCommandWithContext> = []
+
+		const now = this._vmix.getCurrentTime()
+		let nextTimeout = Infinity
+
+		for (const [key, input] of this._inputs.entries()) {
+			if (input.removeTime && input.removeTime <= now) {
+				if (input.added) {
+					commands.push({
+						command: {
+							command: VMixCommand.REMOVE_INPUT,
+							input: input.name,
+						},
+						context: CommandContext.None,
+						timelineId: '',
+					})
+				}
+				this._inputs.delete(key)
+			} else {
+				if (!input.added) {
+					commands.push({
+						command: {
+							command: VMixCommand.ADD_INPUT,
+							value: `${input.type}|${input.name}`,
+						},
+						context: CommandContext.None,
+						timelineId: '',
+					})
+					commands.push({
+						command: {
+							command: VMixCommand.SET_INPUT_NAME,
+							input: this.getFilename(input.name),
+							value: input.name,
+						},
+						context: CommandContext.None,
+						timelineId: '',
+					})
+
+					input.added = true
+				}
+			}
+
+			if (input.removeTime) {
+				nextTimeout = Math.min(nextTimeout, input.removeTime)
+			}
+		}
+
+		this._vmix.addToQueue(commands, now)
+
+		const timeToNextTimeout = nextTimeout - now
+		if (timeToNextTimeout > 0 && timeToNextTimeout !== Infinity) {
+			this._updateTimeout = setTimeout(() => {
```

Comment:
I'm a little confused as to what will be the value of `timeToNextTimeout` when an input gets removed. Say `now := 10000`, and we have two inputs, one to be removed at 10000, another at 10500. Won't `nextTimeout` be set to 10000 and `timeToNextTimeout` to 0, causing a timeout not to be set, resulting in the other input lingering until a third input is scheduled to be removed?

Reply:
Perhaps just a `continue` is needed after [this line](https://github.com/nrkno/sofie-timeline-state-resolver/pull/280/files/39d3ebad6f175fc710bda7a94af6862a6452df80#diff-2bed5085ac736505c2db24ac46af3b4e6682c6f54857a635b6f3837bd8eeb2bfR1515)

Reply:
Good catch! I'll push a fix with a `continue`.